benchmarks_test.go

```diff
@@ -0,0 +1,250 @@
+	}
+	return itms
+}
+
+func MakeSmallPacketDistribution(b *testing.B) [][]byte {
+	b.Helper()
+
+	n := 1000
+	itms := make([][]byte, n)
+	i := 0
+	for ; i < int(float64(n)*(3.0/4.0)); i++ {
+		itms[i] = make([]byte, 64)
+	}
+	for ; i < n; i++ {
+		itms[i] = make([]byte, 1024)
+	}
+	rand.Shuffle(n, func(i, j int) { itms[i], itms[j] = itms[j], itms[i] })
+	return itms
+}
+
+func TestSmallPackets(t *testing.T) {
+	kbps, lat, err := netbench.FindNetworkLimit(testSmallPackets, 0.5)
+	if err != nil {
+		t.Skip()
+	}
+	slowdown, err := netbench.ParallelismSlowdown(testSmallPackets, kbps, lat)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if slowdown > 0.15 {
+		t.Fatalf("Slowdown from mplex was >15%%: %f", slowdown)
+	}
+}
+
+func testSmallPackets(b *testing.B, n1, n2 net.Conn) {
+	msgs := MakeSmallPacketDistribution(b)
+	mpa := NewMultiplex(n1, false)
+	mpb := NewMultiplex(n2, true)
+	mp := runtime.GOMAXPROCS(0)
+	runtime.GOMAXPROCS(mp)
+
+	streamPairs := make([][]*Stream, 0)
+	for i := 0; i < mp; i++ {
+		sa, err := mpa.NewStream()
+		if err != nil {
+			b.Error(err)
+		}
+		sb, err := mpb.Accept()
+		if err != nil {
+			b.Error(err)
+		}
+		streamPairs = append(streamPairs, []*Stream{sa, sb})
+	}
+
+	receivedBytes := uint64(0)
+	sentBytes := uint64(0)
+	idx := int32(0)
+	b.ResetTimer()
+
+	var wg sync.WaitGroup
+
+	b.RunParallel(func(pb *testing.PB) {
+		localIdx := atomic.AddInt32(&idx, 1) - 1
+		localA := streamPairs[localIdx][0]
+		localB := streamPairs[localIdx][1]
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			receiveBuf := make([]byte, 2048)
+
+			for {
+				n, err := localB.Read(receiveBuf)
+				if err != nil && err != io.EOF {
+					b.Error(err)
+				}
+				if n == 0 || err == io.EOF {
+					return
+				}
+				atomic.AddUint64(&receivedBytes, uint64(n))
+			}
+		}()
+
+		i := 0
+		for {
+			n, err := localA.Write(msgs[i])
+			atomic.AddUint64(&sentBytes, uint64(n))
+			if err != nil && err != io.EOF {
+				b.Error(err)
+			}
+			i = (i + 1) % 1000
+			if !pb.Next() {
+				break
+			}
+		}
+		localA.Close()
+	})
+	b.StopTimer()
+	wg.Wait()
+
+	if sentBytes != receivedBytes {
+		b.Fatal("sent != received", sentBytes, receivedBytes)
+	}
+	b.SetBytes(int64(receivedBytes))
+	defer mpa.Close()
+	defer mpb.Close()
+}
+
+func BenchmarkSmallPackets(b *testing.B) {
+	msgs := MakeSmallPacketDistribution(b)
+	benchmarkPackets(b, msgs)
+}
+
+func BenchmarkSlowConnSmallPackets(b *testing.B) {
+	msgs := MakeSmallPacketDistribution(b)
+	slowNetwork := latency.Network{
+		Kbps:    100,
+		Latency: 30 * time.Millisecond,
+		MTU:     1500,
+	}
+	var wg sync.WaitGroup
+	wg.Add(1)
+	var lb net.Conn
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		b.Error(err)
+	}
+	slowL := slowNetwork.Listener(l)
+	go func() {
+		defer wg.Done()
+		lb, _ = slowL.Accept()
+		return
+	}()
+	dialer := slowNetwork.Dialer(net.Dial)
+	la, err := dialer("tcp4", slowL.Addr().String())
+	if err != nil {
+		b.Error(err)
+	}
+	defer la.Close()
```

Comment:
will panic if err != nil

Reply:
Error above should exit already in that case, right?